cpu/rpx0xx/periph/usbdev.c

```diff
@@ -0,0 +1,746 @@
+        &USBCTRL_DPRAM->EP4_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP5_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP6_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP7_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP8_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP9_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP10_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP11_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP12_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP13_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP14_IN_BUFFER_CONTROL,
+        &USBCTRL_DPRAM->EP15_IN_BUFFER_CONTROL,
+    };
+    return reg_map[ep_num];
+}
+
+/* Read the ep-related bit from a status register. */
+static inline bool _reg_get_flag_for_ep(__IOM uint32_t *reg, usbdev_ep_t *ep)
+{
+    uint32_t val = *reg;
+    return val & 1 << (2 * ep->num + (ep->dir == USB_EP_DIR_IN ? 0 : 1));
+}
+
+/* Clear the ep-related bit from a status register. */
+static inline void _reg_clear_flag_for_ep(__IOM uint32_t *reg, usbdev_ep_t *ep)
+{
+    io_reg_atomic_clear(reg, 
+                        1 << (2 * ep->num + (ep->dir == USB_EP_DIR_IN ? 0 : 1)));
+}
+
+static inline __IOM uint32_t * _get_ep_buf_ctrl_reg(uint8_t ep_num,
+                                                    usb_ep_dir_t dir)
+{
+    if (dir == USB_EP_DIR_IN) {
+        return _get_ep_in_buf_ctrl_reg(ep_num);
+    } else {
+        return _get_ep_out_buf_ctrl_reg(ep_num);
+    }
+}
+
+static inline __IOM uint32_t * _get_ep_ctrl_reg(uint8_t ep_num,
+                                                usb_ep_dir_t dir)
+{
+    if (dir == USB_EP_DIR_IN) {
+        return _get_ep_in_ctrl_reg(ep_num);
+    } else {
+        return _get_ep_out_ctrl_reg(ep_num);
+    }
+}
+
+/** See @io_reg_write_dont_corrupt() */
+static inline void _ep_ctrl_reg_write(uint8_t ep_num, usb_ep_dir_t dir,
+                                      uint32_t value, uint32_t mask)
+{
+    __IOM uint32_t *reg = _get_ep_ctrl_reg(ep_num, dir);
+    io_reg_write_dont_corrupt(reg, value, mask);
+}
+
+/** See @io_reg_write_dont_corrupt() */
+static inline void _ep_buf_ctrl_reg_write(uint8_t ep_num, usb_ep_dir_t dir, 
+                                          uint32_t value, uint32_t mask)
+{
+    __IOM uint32_t *reg = _get_ep_buf_ctrl_reg(ep_num, dir);
+    io_reg_write_dont_corrupt(reg, value, mask);
+}
+
+/* Get DPRAM buffer offset. */
+static inline uintptr_t _dpram_offset(uint8_t *buf)
+{
+    return (uintptr_t) buf ^ (uintptr_t) USBCTRL_DPRAM_BASE;
+}
+
+static rpx0xx_usb_ep_t *_get_ep(uint8_t ep_num, usb_ep_dir_t dir)
+{
+    return &_hw_usb_dev.hw_ep[2 * ep_num + (dir == USB_EP_DIR_OUT ? 0 : 1)];
+}
+
+static void _poweron(void)
+{
+    periph_reset(RESETS_RESET_usbctrl_Msk);
+    periph_reset_done(RESETS_RESET_usbctrl_Msk);
+    io_reg_atomic_set(&USBCTRL_REGS->MAIN_CTRL,
+                      USBCTRL_REGS_MAIN_CTRL_CONTROLLER_EN_Msk);
+    io_reg_atomic_set(&USBCTRL_REGS->USB_MUXING,
+                      USBCTRL_REGS_USB_MUXING_TO_PHY_Msk);
```

Comment:
RP2040 datasheet also add `USBCTRL_REGS_USB_MUXING_SOFTCON_Msk`.

+ you should add
```
io_reg_atomic_set(&USBCTRL_REGS->USB_PWR,
                      USBCTRL_REGS_USB_PWR_VBUS_DETECT_Msk
                    | USBCTRL_REGS_USB_PWR_VBUS_DETECT_OVERRIDE_EN_Msk);
```